doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+# The Life of a Connection
+
+This document describes, in broad strokes, how the Conduit proxy works with regard to
+application traffic. We assume that inbound and outbound routing are basically the same,
+modulo a few details that are called out in the proper context.
+
+## Accepting a connection
+
+When an application instance (app) calls a networked service, it resolves the DNS address
+of the service's name to an IP address. It initiates a connection to this address;
+however, because `conduit-proxy-init` has previously configured `iptables(8)`, the
+connection is established to the local `conduit-proxy`.
+
+When the proxy accepts a connection, it first determines the original destination address
+(i.e. the address bound by the client's DNS lookup) via the `SO_ORIGINAL_DST` socket
+option.
+
+It then determines whether there is policy associated with the remote peer's address
+and/or the original destination address. This policy may, for instance, pre-determine how
+the connection should be handled (i.e. with regard to TLS/HTTP/etc).
+
+### Protocol determination
+
+If protocol has not been pre-determined by policy, the proxy attempts to discover the
+protocol by reading data from the client. Note that this detection can break
+server-speaks-first protocols like MySQL and SMTP, necessitating the need to disable such
+detection by policy.
+
+In order to detect the connection's session protocol transparently (i.e. without policy),
+the proxy reads from the connection, buffering data until the proxy determines how to
+handle the connection.
+
+#### TLS
+
+If the connection begins with a TLS[[1]](#tls-q1) client handshake, data is read until
+until a full `ClientHello` message is buffered. The proxy determines whether it should
+terminate the TLS connection or treat the connection as an opaque TCP stream. The proxy
+may terminate TLS when a Conduit-specific extension[[2]](#tls-q2) exists in the client's
+hello, or it may terminate TLS if it is configured with a certificate for the
+client-requested name (i.e. for external/ingress).
+
+If the proxy elects to terminate TLS, the client's ALPN[[3]](#tls-q3) advertisement is
+used to determine which protocol should be used. If the ALPN advertisement does not
+include a proxy-supported protocol (i.e. `http` or `h2`), then the proxy handles the
+decrypted stream as opaque TCP[4](#tls-q4).
+
+##### Questions
+
+<a name="tls-q1"></a>
+1. What's the minimum version of the TLS protocol we support? I assume we don't handle SSL
+    3.0, etc. due to ClientHello requirement
```

Comment:
TLS 1.2, at least for now. Since the only TLS clients we support TLS for initially will be other Conduit clients, we can safely assume that they can do TLS 1.2.